### PR TITLE
Fetch validators from discovery.mligo instead of validators.json

### LINKED ISSUE #160: Validator list: Use discovery.mligo to fetch the URIs of the validators
To keep the PR self contained, #158 uses the existing convention of URIs - `localhost:4440`, `localhost:4441`, `localhost:4442` and so on - for the individual validators. `discovery.mligo` is the source of truth for validator URIs and must be used instead.

### PR DESCRIPTION
**Pull request by anchpop**
## Depends

- [x] #501

## Problem

Currently, the node assumes there will be a `validators.json` file that provides a mapping from validator keys_hashes to URIs. But when new nodes join the network, or when existing nodes want to change their URI, we need some foolproof way to communicate the new URIs to all the nodes. 

## Solution

The discovery.mligo contract was written for this purpose, but before this PR it was unused. Its storage is a map from key_hashes to URIs. Anyone can add or update a key_hash and its associated URI (provided they prove they control the key). After the node fetches the list of validator key_hashes from consensus.mligo, it only has to look up those keys in the map maintained by the discovery.mligo contract.

## Related

- closes  #160

### CODE REVIEW

**Comment by anchpop** [comment]
Marked as a draft because I'm not happy with how inflexible the `Fetch_big_map.run` function is

**Review by d4hines — CHANGES_REQUESTED** [review]
When I tested this, I got errors:
```
Commit operation result: {
  "status": "error",
  "error": "{\"errors\":[{\"kind\":\"temporary\",\"id\":\"proto.011-PtHangz2.michelson_v1.runtime_error\",\"contract_handle\":\"KT1TSZ27g4XAssSs4JvXBCRad9FuC5eKsPCK\",\"contract_code\":[{\"prim\":\""
}
```

Also, please rebase with main.

**Comment by EduardoRFS** [comment]
Please rebase on top of https://github.com/marigold-dev/deku/pull/501

**Comment by aguillon** [comment]
> Please "rebase" on top of #501

@EduardoRFS: I asked @anchpop to propose a not-rebased version because of a bug that he had with your branch. Also, our objectives dictate that this PR is a priority, but @anchpop still has the rebased version as soon as we merge yours.

**Comment by aguillon** [comment]
@EduardoRFS, @d4hines, @anchpop: this PR replaces the validators.json (which were independent for each node) with a common list of validators. I don't think this is what we want for private blockchains. Moreover, I believe that even for a public network we could still have some validators being listed under different addresses depending on the node's configuration (e.g. several private Marigold nodes being relayed by one).

Shouldn't we aim a solution mixing both approaches instead, with identities/addresses listed in the JSON file being prioritized?

**Comment by aguillon** [comment]
@anchpop: your PR looks good to me (see above comment though), but despite trying my best to use Ligo, I can't seem to test this on my machine. Could you please describe what tests you conducted?

**Comment by d4hines** [comment]
@aguillon, there are two separate concerns:

- which public key hashes are allowed to control consensus?
- what domain names are associated with those public key hashes?

Ultimately `consensus.mligo` is the ultimate source of truth for the first question, and that's what makes us a sidechain.

The second question is the same problem of DNS, only it's a map from key hash->domain instead of domain->IP address. Blockchains are useful as DNS's - see ENS and Tezos domains. It's also useful that we can coordinate changes in validator URL's with changes in consensus state.  Lastly, I think it's easier to manage - make a single change to the contract, and all the nodes will update, as opposed to needing to propagate a change to validators.json to every running node.

In short, I disagree with your assertion that a local json file is more useful for private blockchains than a smart contract.

File: src/tezos_interop/tezos_interop.ml [hunk 1]
```diff
@@ -187,16 +204,45 @@ module Consensus = struct
     Listen_transactions.listen t ~rpc_node:t.rpc_node
       ~required_confirmations:t.required_confirmations
       ~destination:t.consensus_contract ~on_message
+  let fetch_discovery t validator_key_hash =
```
[the remark this answers is not in the document]
Comment (d4hines) [reply src/tezos_interop/tezos_interop.ml:207]: fetch_validators*

File: src/tezos_interop/tezos_interop.ml [hunk 2]
```diff
@@ -187,16 +204,45 @@ module Consensus = struct
...
       secret = _;
       bridge_process = _;
     } =
       t in
 
+    let rec traverse_result = function
+      | [] -> Ok []
+      | Ok x :: xs -> Result.bind (traverse_result xs) (fun xs -> Ok (x :: xs))
+      | Error e :: _ -> Error e in
+
```
[the remark this answers is not in the document]
Comment (d4hines) [reply src/tezos_interop/tezos_interop.ml:245]: Woo monad transformers. 

What do we think about adding 
```
List.fold_result : ('a -> 'b -> ('a, 'c) result) -> 'a -> 'b list -> ('a, 'c) result list`
```
and 
```
List.fold_result_lwt : ('a -> 'b -> ('a, 'c) result Lwt.t) -> 'a -> 'b list -> ('a, 'c) result Lwt.t
```
To the `List` module, and then doing it in a single fold_result_lwt down below?

File: src/tezos_interop/tezos_js_bridge.js [hunk 3]
```diff
@@ -23,6 +23,14 @@ const { inspect } = require("util");
...
  * @property {number} confirmation
  * @property {string} destination
+ * 
+ * @typedef BigMapRequest
+ * @type {object}
+ * @property {"storage"} kind
+ * @property {string} rpc_node
+ * @property {number} confirmation
+ * @property {string} destination
+ * @property {any} key
```
[the remark this answers is not in the document]
Comment (d4hines) [reply src/tezos_interop/tezos_js_bridge.js:33]: @EduardoRFS, can you elaborate on why multiple keys would be better? Is the goal to minimize IPC calls in Deku?

File: src/tezos_interop/tezos_interop.mli [hunk 4]
```diff
@@ -34,7 +35,7 @@ module Consensus : sig
     transactions : transaction list;
   }
   val listen_operations : t -> on_operation:(operation -> unit) -> unit
-  val fetch_validators : t -> (Key_hash.t list, string) result Lwt.t
+  val fetch_validators : t -> ((Key_hash.t * Uri.t) list, string) result Lwt.t
```
Comment (d4hines) [inline src/tezos_interop/tezos_interop.mli:38]: Question: should the type of this be `Uri.t option` instead of `Uri.t`? Supposing there's a single key hash that doesn't have an entry in discovery.mligo - do we want to fail the entire node for that?

**Review by d4hines — COMMENTED** [review]
I ran locally and it works for me. The code looks good modulo Edu's comments, but I have a question on what to do with missing URI's.

File: src/tezos_interop/tezos_js_bridge.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (anchpop) [reply src/tezos_interop/tezos_js_bridge.js]: @EduardoRFS I think it's typed properly now but I'm not sure how to check for sure. Also I'm not sure why it would be better to support multiple keys (is the idea that there's a constant overhead when invoking the bridge?)

File: src/tezos_interop/tezos_interop.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (anchpop) [reply src/tezos_interop/tezos_interop.ml]: Both of those are confusing names I think, I renamed it to `fetch_uri_from_discovery`, is that okay?

File: src/tezos_interop/tezos_interop.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (anchpop) [reply src/tezos_interop/tezos_interop.ml]: @EduardoRFS done
@d4hines possibly useful, but since we already have `fold_left_ok` in there, and there's already `Lwt.all`, I'm not sure how useful it would be.

File: src/tezos_interop/tezos_interop.mli (hunk 4 above)
Comment (anchpop) [reply src/tezos_interop/tezos_interop.mli:38]: True, I guess it should be the same as having a node that doesn't respond. But I think that should go in a different PR

**Comment by anchpop** [comment]
@EduardoRFS, @d4hines thanks for those comments, fixed

File: src/tezos_interop/tezos_interop.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (d4hines) [reply src/tezos_interop/tezos_interop.ml]: Sounds good to me.

**Review by d4hines — APPROVED** [review]
LGTM!
Monkey D. Luffy is approves. Time to discover new horizons

File: src/tezos_interop/tezos_js_bridge.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (EduardoRFS) [reply src/tezos_interop/tezos_js_bridge.js]: @d4hines not only to minimize IPC calls, you're also spamming the Tezos node.

[the remark this answers is not in the document]
Comment (EduardoRFS) [reply src/tezos_interop/tezos_js_bridge.js]: it also reduces the complexity and the need for the sequence_result function as now or all fail or none fail.

File: src/tezos_interop/tezos_interop.mli (hunk 4 above)
Comment (EduardoRFS) [reply src/tezos_interop/tezos_interop.mli:38]: @anchpop this should not go in a different PR, this is an unhandled branch.

File: src/tezos_interop/tezos_js_bridge.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (anchpop) [reply src/tezos_interop/tezos_js_bridge.js]: @EduardoRFS I think we'd have to spam the tezos node anyway as I'm not sure there's a way to request the keys  for multiple values in one go.

File: src/tezos_interop/tezos_js_bridge.js [hunk 5]
```diff
@@ -193,6 +201,41 @@ const onStorageRequest = async (id, content) => {
...
+/** @param {BigMapRequest} content */
+const onBigMapRequest = async (id, content) => {
+  const { rpc_node, confirmation, destination, key } = content;
+  const client = new RpcClient(rpc_node);
+  const Tezos = new TezosToolkit(rpc_node);
+  Tezos.setProvider({ config });
+
+  const block = await client.getBlock(); // fetches the head
+
+  const value = await client.getBigMapKey(
```
Comment (EduardoRFS) [inline src/tezos_interop/tezos_js_bridge.js:213]: https://github.com/ecadlabs/taquito/blob/9043379b3d87310892fc9d8ab7c9c14cd8b7a9f4/docs/version.md

This function is deprecated

File: src/tezos_interop/tezos_js_bridge.js [hunk 6]
```diff
@@ -193,6 +201,59 @@ const onStorageRequest = async (id, content) => {
...
+        let index = keys.indexOf(key);
+        if (value !== undefined) {
+          result[index] = value[1];
+        }
+        else {
+          result[index] = null;
+        }
+      });
+      return result;
+    });
```
Comment (EduardoRFS) [inline src/tezos_interop/tezos_js_bridge.js:235]: Please use async await properly, also remove the for loop, it's not clear to me the intention, combine key and values?

File: tests/test_json.ml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+[@@deriving of_yojson { strict = false }, to_yojson]
+
+let () =
+  describe "json" (fun { test; _ } ->
+      test "json" (fun { expect; _ } ->
+          let hmm : success = { values = [Some "hello"; None] } in
+          expect.equal
+            (hmm |> success_to_yojson |> Yojson.Safe.to_string)
+            "\"hello\""))
+*)
```
Comment (EduardoRFS) [inline tests/test_json.ml:54]: What about all this commented code?

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

**Review by EduardoRFS — APPROVED** [review]
Please, just test, and if it's fully functional we can merge. My bad I had introduced a bug yesterday.

anime girl smiling to this PR
